invokeai/app/services/download/download_default.py

```diff
@@ -0,0 +1,394 @@
+# Copyright (c) 2023, Lincoln D. Stein
+"""Implementation of multithreaded download queue for invokeai."""
+
+import os
+import re
+import threading
+import traceback
+from logging import Logger
+from pathlib import Path
+from queue import PriorityQueue
+from typing import Any, Dict, List, Optional, Set
+
+import requests
+from pydantic.networks import AnyHttpUrl, Url
+from requests import HTTPError
+from tqdm import tqdm
+
+from invokeai.app.services.events.events_base import EventServiceBase
+from invokeai.app.util.misc import get_iso_timestamp
+from invokeai.backend.util.logging import InvokeAILogger
+
+from .download_base import (
+    DownloadEventHandler,
+    DownloadJob,
+    DownloadJobCancelledException,
+    DownloadJobStatus,
+    DownloadQueueServiceBase,
+    UnknownJobIDException,
+)
+
+# Maximum number of bytes to download during each call to requests.iter_content()
+DOWNLOAD_CHUNK_SIZE = 100000
+
+# marker that the queue is done and that thread should exit
+STOP_JOB = DownloadJob(id=-99, priority=-99, source=Url("http://dummy.com/"), dest=Path("/"))
+
+
+class DownloadQueueService(DownloadQueueServiceBase):
+    """Class for queued download of models."""
+
+    _jobs: Dict[int, DownloadJob]
+    _max_parallel_dl: int = 5
+    _worker_pool: Set[threading.Thread]
+    _queue: PriorityQueue
+    _lock: threading.Lock
+    _logger: Logger
+    _events: Optional[EventServiceBase] = None
+    _next_job_id: int = 0
+    _requests: requests.sessions.Session
+
+    def __init__(
+        self,
+        max_parallel_dl: int = 5,
+        event_bus: Optional[EventServiceBase] = None,
+        requests_session: Optional[requests.sessions.Session] = None,
+    ):
+        """
+        Initialize DownloadQueue.
+
+        :param max_parallel_dl: Number of simultaneous downloads allowed [5].
+        :param requests_session: Optional requests.sessions.Session object, for unit tests.
+        """
+        self._jobs = {}
+        self._next_job_id = 0
+        self._queue = PriorityQueue()
+        self._worker_pool = set()
+        self._lock = threading.Lock()
+        self._logger = InvokeAILogger.get_logger("DownloadQueueService")
+        self._event_bus = event_bus
+        self._requests = requests_session or requests.Session()
+        self._max_parallel_dl = max_parallel_dl
+
+    def start(self, *args: Any, **kwargs: Any) -> None:
+        """Start the download worker threads."""
+        self._start_workers(self._max_parallel_dl)
+
+    def stop(self, *args: Any, **kwargs: Any) -> None:
+        """Stop the download worker threads."""
+        for thread in self._worker_pool:
+            if thread.is_alive():
+                self._queue.put(STOP_JOB)
```

Comment:
I think there's a race condition here. Consider the following sequence:
1. The main thread checks `self._worker_pool[0]`. It's alive, so it puts a `STOP_JOB` on the queue.
2. The worker thread at `self._worker_pool[1]` processes the `STOP_JOB` and terminates.
3. The main thread checks `self._worker_pool[1]`. It's not alive. Now the main thread won't put enough `STOP_JOB`s on the queue.

Could solve by counting alive threads first and then adding that number of `STOP_JOB`s to the queue. But, this opens the door for a thread to fail for some other reason and then there will always be a `STOP_JOB` left in the queue, meaning that `join(...)` would hang forever.

For this reason, I think the solution discussed here would be more robust: https://github.com/invoke-ai/InvokeAI/pull/5225#discussion_r1428651188

Reply:
Hadn't thought of that, and this convinced me that the `threading.Event` solution was better.